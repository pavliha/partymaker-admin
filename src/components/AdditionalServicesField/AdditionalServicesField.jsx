import React, { Component } from 'react'
import { object, func, bool, string, arrayOf, shape } from 'prop-types'
import { withStyles } from '@material-ui/styles'
import serviceShape from 'shapes/service'
import { IconButton, FormHelperText, Input } from '@material-ui/core'
import MinusCircleOutlineIcon from 'mdi-react/MinusCircleOutlineIcon'
import AddCircleOutlineIcon from 'mdi-react/AddCircleOutlineIcon'
import { NumberField } from 'components'
import uniqId from 'uniqid'
import { actions, connect } from 'src/redux'

const styles = theme => ({
  root: {
    width: '100%',
  },

  table: {
    borderCollapse: 'collapse',
    width: '100%',

    '& th': {
      textAlign: 'left',
      fontFamily: 'Google Sans,sans-serif',
      fontWeight: 400,
      paddingBottom: 15,
    },
    '& td': {
      borderTop: '1px rgba(0,0,0,0.1) solid',
      borderBottom: '1px rgba(0,0,0,0.1) solid',

    }
  },

  iconButton: {
    padding: 5,
  },

  titleInput: {
    fontFamily: 'Google Sans,sans-serif',
    fontWeight: 400,
    color: theme.palette.secondary.main,
    fontSize: 16,
  },

  descriptionInput: {
    fontSize: 14,
    color: 'rgba(0,0,0,0.5)'
  }
})

class AdditionalServicesField extends Component {

  state = {
    title: '',
    price: null,
    description: '',
  }

  handleTitle = e => {
    this.setState({ title: e.target.value })
  }

  handleDescription = e => {
    this.setState({ description: e.target.value })
  }

  handlePrice = (name, value) => {
    this.setState({ price: value })
  }

  changeTitle = index => e => {
    const { name, value: services, onChange } = this.props
    services[index] = { ...services[index], title: e.target.value }
    onChange(name, [...services])
  }

  changeDescription = index => e => {
    const { name, value: services, onChange } = this.props
    services[index] = { ...services[index], description: e.target.value }
    onChange(name, [...services])
  }

  changePrice = index => (field, value) => {
    const { name, value: services, onChange } = this.props
    services[index] = { ...services[index], price: value }
    onChange(name, [...services])
  }

  add = () => {
    const { name, value: services, onChange } = this.props
    const { title, price, description } = this.state
    const service = { id: uniqId(), title, price, description }
    onChange(name, [...services, service])
    this.setState({ title: '', price: '', description: '' })
  }

  remove = (index) => () => {
    const { name, value: services, onChange, redux } = this.props
    redux.remove(services[index].id)
    services.splice(index, 1)
    onChange(name, [...services])
  }

  render() {
    const { classes, value: services, error, helperText } = this.props
    const { title, price, description } = this.state

    return (
      <div className={classes.root}>
        <table className={classes.table}>
          <thead>
            <tr>
              <th>Услуга</th>
              <th>Цена</th>
            </tr>
          </thead>
          <tbody>
            {services.map((service, index) => (
              <tr key={service.id}>
                <td>
                  <Input
                    className={classes.titleInput}
                    fullWidth
                    disableUnderline
                    value={service.title}
                    placeholder="Назавание услуги"
                    onChange={this.changeTitle(index)}
                  />
                  <Input
                    className={classes.descriptionInput}
                    fullWidth
                    disableUnderline
                    value={service.description}
                    placeholder="Описание услуги"
                    onChange={this.changeDescription(index)}
                  />
                </td>
                <td>
                  <NumberField
                    InputProps={{ disableUnderline: true }}
                    margin="normal"
                    value={service.price}
                    suffix=" грн"
                    placeholder="100 грн"
                    onChange={this.changePrice(index)}
                  />
                </td>
                <td>
                  <IconButton
                    color="secondary"
                    className={classes.iconButton}
                    onClick={this.remove(index)}
                  >
                    <MinusCircleOutlineIcon />
                  </IconButton>
                </td>
              </tr>
            ))}
            <tr>
              <td>
                <Input
                  fullWidth
                  className={classes.titleInput}
                  value={title}
                  disableUnderline
                  placeholder="Назавание услуги"
                  onChange={this.handleTitle}
                />
                <Input
                  className={classes.descriptionInput}
                  fullWidth
                  value={description}
                  disableUnderline
                  placeholder="Описание услуги"
                  onChange={this.handleDescription}
                />
              </td>
              <td width="100px">
                <NumberField
                  value={price}
                  suffix=" грн"
                  InputProps={{ disableUnderline: true }}
                  placeholder="100 грн"
                  onChange={this.handlePrice}
                />
              </td>
              <td width="25px">
                <IconButton
                  color="secondary"
                  className={classes.iconButton}
                  disabled={!title}
                  onClick={this.add}
                >
                  <AddCircleOutlineIcon />
                </IconButton>
              </td>
            </tr>
          </tbody>
        </table>
        <FormHelperText error={error}>{helperText}</FormHelperText>
      </div>
    )
  }
}

AdditionalServicesField.propTypes = {
  classes: object.isRequired,
  name: string,
  value: arrayOf(shape(serviceShape)),
  helperText: string,
  error: bool,
  onChange: func.isRequired,
  redux: object,
}

const redux = () => ({
  remove: actions.places.additional_services.remove
})
export default withStyles(styles)(connect(redux)(AdditionalServicesField))
