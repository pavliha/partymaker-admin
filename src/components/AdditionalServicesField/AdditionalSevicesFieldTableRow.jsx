import React from 'react'
import { IconButton, Input } from '@material-ui/core'
import { withStyles } from '@material-ui/styles'
import { NumberField } from 'components'
import { number, object, oneOfType, shape, string, func } from 'prop-types'
import MinusCircleOutlineIcon from 'mdi-react/MinusCircleOutlineIcon'

const styles = theme => ({

  iconButton: {
    padding: 5,
  },

  titleInput: {
    fontFamily: 'Google Sans,sans-serif',
    fontWeight: 400,
    color: theme.palette.secondary.main,
    fontSize: 16,
  },

  descriptionInput: {
    fontSize: 14,
    color: 'rgba(0,0,0,0.5)'
  }
})

const AdditionalServicesFieldTableRow = ({ classes, service, onChange, onDelete }) =>
  <tr data-testid="AdditionalServicesField-tr">
    <td>
      <Input
        className={classes.titleInput}
        fullWidth
        disableUnderline
        data-testid="AdditionalServicesField-title"
        value={service.title}
        placeholder="Назавание услуги"
        onChange={e => onChange({ ...service, title: e.target.value })}
      />
      <Input
        className={classes.descriptionInput}
        fullWidth
        disableUnderline
        value={service.description}
        data-testid="AdditionalServicesField-description"
        placeholder="Описание услуги"
        onChange={e => onChange({ ...service, description: e.target.value })}
      />
    </td>
    <td>
      <NumberField
        InputProps={{ disableUnderline: true }}
        margin="normal"
        value={service.price}
        suffix=" грн"
        data-testid="AdditionalServicesField-price"
        placeholder="100 грн"
        onChange={value => onChange({ ...service, price: value })}
      />
    </td>
    <td>
      <IconButton
        color="secondary"
        className={classes.iconButton}
        data-testid="AdditionalServicesField-remove"
        onClick={() => onDelete(service)}
      >
        <MinusCircleOutlineIcon />
      </IconButton>
    </td>
  </tr>

AdditionalServicesFieldTableRow.propTypes = {
  classes: object.isRequired,
  service: shape({
    id: oneOfType([string, number]).isRequired,
    title: string,
    description: string
  }),
  onChange: func.isRequired,
  onDelete: func.isRequired,
}
export default withStyles(styles)(AdditionalServicesFieldTableRow)
