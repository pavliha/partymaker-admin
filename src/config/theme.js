export default {
  typography: {
    useNextVariants: true,
  },

  palette: {
    primary: {
      light: '#D404DC',
      main: '#9306BC',
      dark: '#9306BC',
    },
    error: {
      main: '#D6000A',
    },
  },

  overrides: {
    MuiTableCell: {
      root: {
        '&$paddingNone': {
          '&last-child': {
            padding: 'inherit'
          }
        }
      },
      paddingNone: {
        '&last-child': {
          padding: 'inherit'
        }
      }
    }
  }
}
