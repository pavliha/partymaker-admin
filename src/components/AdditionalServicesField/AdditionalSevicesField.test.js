import React from 'react'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import AdditionalServicesField from './AdditionalServicesField'
import theme from 'config/theme'
import { ThemeProvider } from '@material-ui/styles'
import { createMuiTheme } from '@material-ui/core'

describe('AdditionalServicesField', () => {

  it('renders with minimal props', () => {
    const mockStore = configureStore([])
    const store = mockStore({})

    const additional_services = [
      {
        id: 1,
        title: 'Аренда беседок',
        description: 'Вмешает от 1 до 6 человек',
        price: 100,
      }, {
        id: 2,
        title: 'Аренда беседок',
        description: 'Вмешает от 1 до 6 человек',
        price: 100,
      }
    ]

    const tree = renderer
      .create(
        <ThemeProvider theme={createMuiTheme(theme)}>
          <Provider store={store}>
            <AdditionalServicesField
              name="additional_services"
              value={additional_services}
              onChange={() => {}}
            />
          </Provider>
        </ThemeProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

})
