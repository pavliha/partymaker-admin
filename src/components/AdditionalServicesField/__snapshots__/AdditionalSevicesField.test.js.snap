// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with minimal props 1`] = `
<div
  className="Connect(AdditionalServicesField)-root-1"
>
  <table
    className="Connect(AdditionalServicesField)-table-2"
  >
    <thead>
      <tr>
        <th>
          Услуга
        </th>
        <th>
          Цена
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <div
            className="MuiInputBase-root MuiInput-root Connect(AdditionalServicesField)-titleInput-4 MuiInputBase-fullWidth MuiInput-fullWidth"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Назавание услуги"
              type="text"
              value="Аренда беседок"
            />
          </div>
          <div
            className="MuiInputBase-root MuiInput-root Connect(AdditionalServicesField)-descriptionInput-5 MuiInputBase-fullWidth MuiInput-fullWidth"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Описание услуги"
              type="text"
              value="Вмешает от 1 до 6 человек"
            />
          </div>
        </td>
        <td>
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
            onKeyDown={[Function]}
            onMouseUp={[Function]}
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="100 грн"
                required={false}
                type="text"
                value="100 грн"
              />
            </div>
          </div>
        </td>
        <td>
          <button
            className="MuiButtonBase-root MuiIconButton-root Connect(AdditionalServicesField)-iconButton-3 MuiIconButton-colorSecondary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                className="mdi-icon "
                fill="currentColor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2M7,13H17V11H7"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="MuiInputBase-root MuiInput-root Connect(AdditionalServicesField)-titleInput-4 MuiInputBase-fullWidth MuiInput-fullWidth"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Назавание услуги"
              type="text"
              value="Аренда беседок"
            />
          </div>
          <div
            className="MuiInputBase-root MuiInput-root Connect(AdditionalServicesField)-descriptionInput-5 MuiInputBase-fullWidth MuiInput-fullWidth"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Описание услуги"
              type="text"
              value="Вмешает от 1 до 6 человек"
            />
          </div>
        </td>
        <td>
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
            onKeyDown={[Function]}
            onMouseUp={[Function]}
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="100 грн"
                required={false}
                type="text"
                value="100 грн"
              />
            </div>
          </div>
        </td>
        <td>
          <button
            className="MuiButtonBase-root MuiIconButton-root Connect(AdditionalServicesField)-iconButton-3 MuiIconButton-colorSecondary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                className="mdi-icon "
                fill="currentColor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2M7,13H17V11H7"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="MuiInputBase-root MuiInput-root Connect(AdditionalServicesField)-titleInput-4 MuiInputBase-fullWidth MuiInput-fullWidth"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Назавание услуги"
              type="text"
              value=""
            />
          </div>
          <div
            className="MuiInputBase-root MuiInput-root Connect(AdditionalServicesField)-descriptionInput-5 MuiInputBase-fullWidth MuiInput-fullWidth"
            onClick={[Function]}
          >
            <input
              className="MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Описание услуги"
              type="text"
              value=""
            />
          </div>
        </td>
        <td
          width="100px"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
            onKeyDown={[Function]}
            onMouseUp={[Function]}
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="100 грн"
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
        </td>
        <td
          width="25px"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root Connect(AdditionalServicesField)-iconButton-3 MuiIconButton-colorSecondary Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                className="mdi-icon "
                fill="currentColor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2M13,7H11V11H7V13H11V17H13V13H17V11H13V7Z"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <p
    className="MuiFormHelperText-root"
  />
</div>
`;
